File: classes-ObjC/STCardsMenu/STCardsMenuAnimation.m
//
//  STCardsMenuAnimation.m
//  STCardsMenu-ObjC
//
//  Created by TangJR on 12/14/15.
//  Copyright © 2015 tangjr. All rights reserved.
//

#import "STCardsMenuAnimation.h"
#import "STCardsMenuConst.h"

static const NSTimeInterval STCardsMenuAnimationDuration = 0.5;

@implementation STCardsMenuAnimation

+ (void)cardIdentityAnimationWithView:(UIView *)view {
    view.userInteractionEnabled = NO;
    [UIView animateWithDuration:STCardsMenuAnimationDuration animations:^{
        view.transform = CGAffineTransformIdentity;
    } completion:^(BOOL finished) {
        view.layer.cornerRadius = 0.0;
        view.userInteractionEnabled = YES;
    }];
}

+ (void)cardScaleLittleAnimationWithView:(UIView *)view point:(CGPoint)point {
    view.userInteractionEnabled = NO;
    [UIView animateWithDuration:STCardsMenuAnimationDuration animations:^{
        view.transform = CGAffineTransformMakeTranslation(point.x, point.y);
        view.layer.cornerRadius = 8.0;
    } completion:^(BOOL finished) {
        view.userInteractionEnabled = YES;
    }];
}

+ (void)cardMoveToRightAnimationWithView:(UIView *)view {
    view.userInteractionEnabled = NO;
    [UIView animateWithDuration:STCardsMenuAnimationDuration animations:^{
        view.transform = CGAffineTransformMakeTranslation(CGRectGetWidth([UIScreen mainScreen].bounds), CGRectGetMinY(view.frame));
    } completion:^(BOOL finished) {
        view.layer.cornerRadius = 0.0;
        view.userInteractionEnabled = YES;
    }];
}

+ (void)titleIdentityAnimationWithView:(UIView *)view {
    [UIView animateWithDuration:STCardsMenuAnimationDuration animations:^{
        view.transform = CGAffineTransformIdentity;
    }];
}

+ (void)titleMoveToRightAnimationWithView:(UIView *)view {
    [UIView animateWithDuration:STCardsMenuAnimationDuration animations:^{
        view.transform = CGAffineTransformMakeTranslation(CGRectGetMinX(view.frame) + 40, CGRectGetMinY(view.frame));
    }];
}

+ (void)hamburgIdentityAnimationWithView:(UIView *)view {
    [UIView animateWithDuration:STCardsMenuAnimationDuration animations:^{
        view.transform = CGAffineTransformIdentity;
        view.alpha = 0;
    }];
}

+ (void)hamburgShowAnimationWithView:(UIView *)view {
    [UIView animateWithDuration:STCardsMenuAnimationDuration animations:^{
        view.alpha = 1;
        view.transform = CGAffineTransformMakeTranslation(20, CGRectGetMinY(view.frame));
    }];
}

+ (void)closeButtonShowAnimationWithView:(UIView *)view {
    view.hidden = NO;
    [UIView animateWithDuration:STCardsMenuAnimationDuration animations:^{
        view.alpha = 1;
        view.transform = CGAffineTransformIdentity;
    }];
}

+ (void)closeButtonHideAnimationWithView:(UIView *)view {
    [UIView animateWithDuration:STCardsMenuAnimationDuration animations:^{
        view.alpha = 0;
        view.transform = CGAffineTransformConcat(CGAffineTransformMakeRotation(M_PI_4 / 4), CGAffineTransformMakeTranslation(STVCX - 5.0, STVCY + 5.0));
    } completion:^(BOOL finished) {
        view.hidden = YES;
    }];
}

@end

File: classes-Swift/STCardsMenu/STCardsMenuAnimation.swift
//
//  File.swift
//  STCardsMenu-Swift
//
//  Created by TangJR on 12/16/15.
//  Copyright © 2015 tangjr. All rights reserved.
//

import UIKit

class STCardsMenuAnimation {
    static func cardsMenuIdentityAnimation(view: UIView) {
        UIView.animateWithDuration(STCardsMenuConst.AnimationDuration, animations: { () -> Void in
            view.transform = CGAffineTransformIdentity
            })  { (finish: Bool) -> Void in
                view.layer.cornerRadius = 0.0
        }
    }
    
    static func cardsMenuToRightAnimation(view: UIView) {
        UIView.animateWithDuration(STCardsMenuConst.AnimationDuration, animations: { () -> Void in
            view.transform = CGAffineTransformMakeTranslation(CGRectGetWidth(UIScreen.mainScreen().bounds), CGRectGetMinY(view.frame))
            }) { (finish: Bool) -> Void in
                view.layer.cornerRadius = 0.0
        }
    }
    
    static func cardsMenuScaleLittleAnimation(view: UIView, point: CGPoint) {
        UIView.animateWithDuration(STCardsMenuConst.AnimationDuration, animations: { () -> Void in
            view.transform = CGAffineTransformMakeTranslation(point.x, point.y)
            view.layer.cornerRadius = 8.0
            }) { (finish: Bool) -> Void in
                
        }
    }
    
    static func hamburgButtonHideAnimation(hamburgButton: STCardsMenuHamburgButton, titleLabel: UILabel) {
        hamburgButton.hidden = false
        hamburgButton.alpha = 1.0
        UIView.animateWithDuration(STCardsMenuConst.AnimationDuration, animations: { () -> Void in
            hamburgButton.transform = CGAffineTransformIdentity
            hamburgButton.alpha = 0.0
            titleLabel.transform = CGAffineTransformIdentity
            }) { (finish: Bool) -> Void in
                hamburgButton.hidden = true
        }
    }
    
    static func hamburgButtonShowAnimation(hamburgButton: STCardsMenuHamburgButton, titleLabel: UILabel) {
        hamburgButton.hidden = false
        hamburgButton.alpha = 0.0
        UIView.animateWithDuration(STCardsMenuConst.AnimationDuration, animations: { () -> Void in
            titleLabel.transform = CGAffineTransformMakeTranslation(CGRectGetMaxX(hamburgButton.frame), CGRectGetMinY(titleLabel.frame))
            hamburgButton.alpha = 1.0
            }) { (finish: Bool) -> Void in
                
        }
    }
    
    static func closeButtonHideAnimation(closeButton: STCardsMenuCloseButton) {
        UIView.animateWithDuration(STCardsMenuConst.AnimationDuration, animations: { () -> Void in
            closeButton.alpha = 0.0
            closeButton.transform = CGAffineTransformConcat(CGAffineTransformMakeRotation(CGFloat(M_PI_4) / 4.0), CGAffineTransformMakeTranslation(STCardsMenuConst.VCX - 5.0, STCardsMenuConst.VCY + 5.0))
            }) { (finish: Bool) -> Void in
                closeButton.hidden = true
        }
    }
    
    static func closeButtonShowAnimation(closeButton: STCardsMenuCloseButton) {
        closeButton.hidden = false
        UIView.animateWithDuration(STCardsMenuConst.AnimationDuration, animations: { () -> Void in
            closeButton.alpha = 1.0
            closeButton.transform = CGAffineTransformIdentity
            }) { (finish: Bool) -> Void in
                
        }
    }
}